ENH: cam-specific controls

ZLLentz opened the pull request:
Request from UED
Awaiting feedback in https://jira.slac.stanford.edu/browse/ECS-7005 but I think the code is done now.

The main idea here is that UED wanted camera-specific controls for their Andor cameras because they need to change the gain a lot and using two GUIs is burdensome to operations.

Every cam will get a camera controls section on the main screen with at least start/stop controls, and specific camera models can be granted extra behavior.

This is implemented by adding an additional empty QGroupBox, giving it a QFormLayout, and repeatedly emptying and filling this layout as needed when cameras swap around.

I implemented this in a submodule and had it clean itself up on camera switches instead of trying to weave it into the big camviewer file. I didn't use any ui files because I wanted to keep this simple and limited.

![image](https://github.com/user-attachments/assets/c2327049-5fec-46bf-9d57-7c98f8f7073e)

ZLLentz committed:
WIP: start implementing model specific screens

ZLLentz committed:
ENH: more or less get the model-specific controls to work

ZLLentz committed:
FIX: higher maximum value for em gain, only put after editing is done

ZLLentz committed:
FIX: ensure cleanup of cam-specific screen at app close

tangkong commented on models.py:
Perhaps a dangerously overloaded term in the model-view-controller world qt lives in.  But maybe that's just my biases talking

ZLLentz replied:
You are absolutely right here. This should be renamed to something less overloaded.
Maybe cam_types.py?

tangkong replied:
I'm happy with cam_types.py 👍

tangkong commented on models.py:
```diff
@@ -0,0 +1,205 @@
+        with self.finish_lock:
+            if self.finish_ran:
+                return
+            self.finish_ran = True
```
If we encounter an error anywhere in the body of this, we'll be locked out from ever being able to `.finish_form()` again.  Is this what you intended?  I suppose we're ok with this since people aren't really interacting with these methods

ZLLentz replied:
I was much more concerned with accidentally running this twice than I was with the form failing to complete.

ZLLentz replied:
Maybe this was not correct

tangkong replied:
I think it's a valid concern, and one that you addressed rather thoroughly.  Perhaps reading this from the point of view of `ModelScreenGenerator`, I was just inherently distrustful of the downstream functions

ZLLentz replied:
I ended up wrapping the downstream functions a bit to make sure they raise a warning pop-up instead of crashing the app

tangkong commented on models.py:
```diff
@@ -0,0 +1,205 @@
+            ...
+
+    gain_spinbox = QSpinBox()
+    gain_spinbox.setMaximum(10000000)
```
that's a lot of gain

ZLLentz replied:
I have no idea what the actual correct number is and the jira is pretty silent

tangkong replied:
Give them all the gain!  I imagine we should probably be grabbing a max value from the PV metadata, but that can't be trusted 😢

ZLLentz replied:
the metadata says the limits are 0 and 0
I could go to the manual I guess

tangkong replied:
That's on the IOC creator :/

ZLLentz replied:
Fwiw I couldn't find a manual that put a bound on this value

tangkong reviewed (commented):
This seems reasonable to me, I'll take it for a spin before stamping.  

Almost done with camviewer huh? 😆

tangkong commented:
Apparently the UED camera IOCs are down so I can't test these on the UED cameras in question, but I do think it's nice that we can see a more fully qualified name for the camera in camviewer now.

ZLLentz commented:
UED cam IOCs are only available from the UED subnet, not even to psbuild.

tangkong commented:
ah right that special subnet.  Silly me

tangkong commented:
I think I managed to break it by selecting a camera that was indeed down, then trying to swiftly switch back to one of the Andors.  

Here I have a gige selected but the andor specific controls are there, and the entire UI is frozen

<img width="290" alt="image" src="https://github.com/user-attachments/assets/0ee2b1f6-26a9-487c-bed7-9412d10ad8b3" />

tangkong commented:
Which server are you testing this on?  I just picked a UED server but maybe I'm still on the wrong subnet

ZLLentz commented:
> Here I have a gige selected but the andor specific controls are there, and the entire UI is frozen

And dangerously, the start/stop buttons are still visible from the previous camera... I'll have to add an additional clear step.

> Which server are you testing this on?

I was using `ued-daq` and being really careful not to actually change their settings. I was also running on `tst-console` with `/cds/home/z/zlentz/camviewer.cfg` with a silly temporary IOC hooked up to a camera in b901.

tangkong commented:
Interesting, that's where I ended up too  (ued-daq).

Another thing to point out is how the spinbox didn't get initialized to the same value as the gain number?

ZLLentz commented:
>  the spinbox didn't get initialized to the same value as the gain number?

I forgot to do this, I will add it

ZLLentz committed:
MNT: rename model to cam types to avoid name collision with MVC model

ZLLentz committed:
FIX/MNT: update spinbox too when em gain changes, reorganize gain display code

ZLLentz committed:
ENH: error message instead of crash if we break a model screen

ZLLentz committed:
FIX: include model screen generator in cleanup to avoid desync

ZLLentz committed:
MNT: update attribute name to match related renames

ZLLentz committed:
MNT: only need to clean up once

ZLLentz commented:
I think I addressed the items here but I will spend some time later making sure I didn't compromise the user experience. The way I've implemented the more thorough cleanup is a bit inconsistent, for example.

ZLLentz committed:
MNT: more thorough cleanup and consistent visuals for not connected yet

ZLLentz committed:
ENH: enable and disable the acquire buttons with write access

ZLLentz committed:
ENH: handle write access for EM gain too

ZLLentz committed:
MNT: remove unused get_layout

ZLLentz committed:
DOC: add docstrings for cam_types module

ZLLentz committed:
MNT: rename for clarity

ZLLentz committed:
ENH/DOC: protect the user from a poorly timed em config, expand doc

ZLLentz commented:
- Responded to review suggestions
- Added docstrings
- Cleaned up the UX for not-loading cameras
- Disabled the run widgets when they wouldn't work (e.g. no cam loaded, no write access)
- Disabled the EM Gain widgets when the cam is running (UED says cam running + gain change = dead cam)

tangkong commented on camviewer_ui_impl.py:
```diff
@@ -2026,6 +2038,21 @@ def connectCamera(self, sCameraPv, index, sNotifyPv=None):
+        else:
+            self.cam_type_screen_generator.cleanup()
+            form = self.ui.groupBoxControls.layout()
+        self.cam_type_screen_generator = CamTypeScreenGenerator(self.ctrlBase, form)
```
So every time we set a new camera, we create a new `CamTypeScreenGenerator`?  Do we ever unparent the old ones or deleteLater them? 

In flipping between cameras I never ran into a problem/memory leak, but I am suspicious.  It seems like we're already disconnecting signals and removing rows from the form, so could we just repopulate it?

Ultimately I think I'm ok with this as is.  If we run into performance issues we can revisit.

ZLLentz replied:
You are correct, replacing this object should no longer be necessary with some small tweaks. (I won't fix this now, in favor of forward progress through the last few camviewer tickets)

In an earlier version of this PR I was trying to remove the `QFormLayout` from the `QGroupBox` which actually seems to be impossible in pyqt. In that earlier version the screen generator was a one-time use object that had custody of a unique layout per screen generator object, so I was removing and replacing them. It was only later that I switched to clearing and re-using the layout instead.

tangkong reviewed (approved):
The problems I ran into in my previous tests have been resolved 🎉 

Looks like things get cleaned up nicely as well.  I'm onboard  👍
